add blog post for constructors in golang (#213)
deliveroo/deliveroo.engineering: _posts/2020-09-04-increase-the-reliability-of-a-golang-codebase-with-object-constructors.md

---
layout: post
title:  "Increase the Reliability of a Golang Codebase with Object Constructors"
authors:
  - "Tugberk Ugurlu"
excerpt: >

>   One of the limitation of Go programming language is the lack of built-in object constructor 

Reviewer: limitations (because there are many)

>   One of the limitation of Go programming language is the lack of built-in object constructor 

Author: ```suggestion
  One of the limitations of Go programming language is the lack of built-in object constructor 
```

  support. In this post, we will see how this can have a negative impact on the code we write 
  and how we can get around that by gluing together some of the existing language concepts.
---

Coming from a heavy production experience with languages such as C# and TypeScript, I must admit 
that my journey with Go has been a bumpy ride so far, even if the scale is definitely heavier on 

> the positive side. Go certainly shines in some parts such as its runtime efficiency, built-in 

Reviewer: I'm not really sure that "even if the scale is definitely heavier on 
the positive side" makes much sense - what are you trying to say?

tooling support and its simplicity which allows you to get up to speed with it so quickly! 
However, there are some areas where it limits your ability to express and model your 
software in code in a robust way, especially in a codebase where you get to work on as a team 
such as lack of [sum types](https://github.com/golang/go/issues/19412) and generics support (lucky, [generics support seems to be on its way](https://go.googlesource.com/proposal/+/master/design/go2draft-generics-overview.md)). 
One of these limitations I have come across is not [having any built-in constructor support](https://twitter.com/tourismgeek/status/1074325233220374528).

> I came across this limitation before using Go in an anger and I was anticipating this to be 

Reviewer 2: I would change or remove this bit.

> I came across this limitation before using Go in an anger and I was anticipating this to be a bit of a limitation

Author: changed https://github.com/deliveroo/deliveroo.engineering/pull/213/commits/d9639dd88867a4bf190bf2be0dcd1e4da7d08634

> I came across this limitation before using Golang in an anger and I was anticipating this to be 

Reviewer: > I came across this limitation before using Golang in an anger and I was anticipating this to be 
a bit of a limitation. At the same time, I was also being open-minded. 

I think this could be rewritten

Author: done in https://github.com/deliveroo/deliveroo.engineering/pull/213/commits/d9639dd88867a4bf190bf2be0dcd1e4da7d08634

mentioning and I will list them below.

### Eliminates the risk of being unable to act on implementation struct having new mandatory fields

This is because the compiler would not help us here if we were to construct the struct through struct initialization and start assigning the fields directly as Go doesn't have a way to enforce this unlike other languages such as TypeScript do. We need to follow all the usage and update them manually, which is going to be tedious and error-prone in a large codebase. At best case scenario, you will have some unit tests broken if they are written correctly. In a worst case scenario, your code will blow up at Runtime which will require you to rollback your release. Believe it or not, this is still not the worst case, The worst case here is that your application will happily work w/o any crashes but the behaviour of the application could be wrong due to the way the implementation might work. This one is the hardest to catch and more painful one to swallow as it could have a larger impact on your efforts and outcome you wanted to achieve in the first place.

> Let's imagine a case for our jazzSinger to start getting the lyrics from an external resource. You would structure this by providing an interface and allowing jazzSinger to call into that, which may look like below:

Reviewer 3: Let's imagine our jazzSinger would start getting lyrics from an external resource. ....

... which would look like the following snippet/example:

```golang
package singers

// Lyrics

type LyricsProvider interface {
	GetRandom() string
}

type jazzLyricsProvider struct {
}

func (jazzLyricsProvider) GetRandom() string {
	return "Des yeux qui font baisser les miens"
}

func NewJazzLyricsProvider() LyricsProvider {
	return jazzLyricsProvider{}
}

// Singer

type Singer interface {
	Sing() string
}

type jazzSinger struct {
	lyrics LyricsProvider
}

func (js jazzSinger) Sing() string {
	return js.lyrics.GetRandom() 
}

func NewJazzSinger(lyrics LyricsProvider) Singer {
	return jazzSinger{
		lyrics: lyrics,
	}
}
```

> If we were to build our application directly without modifying the main package (which is the consumer of the singers package), we will receive the below error:

Reviewer 3: ... we would see the following error:

```
➜  go-package-scope go build main.go 
# command-line-arguments
./main.go:9:28: not enough arguments in call to singers.NewJazzSinger
	have ()
	want (singers.LyricsProvider)
```

> We wouldn't get this level of feedback if we were to directly initialize the struct directly. What we wouldn't get instead is a failure on runtime:

Reviewer 3: ... What we would get instead is a failure...

```
➜  go-package-scope go run main.go  
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1091512]

goroutine 1 [running]:
github.com/tugberkugurlu/go-package-scope/singers.JazzSinger.Sing(0x0, 0x0, 0x1010095, 0xc00000e1e0)
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/singers/jazzsinger.go:31 +0x22
main.singToConsole(0x10d7520, 0xc00000e1e0)
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/main.go:14 +0x35
main.main()
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/main.go:10 +0x57
exit status 2
```

### Allows you to provide parameter validation as early as possible 

> Which also allows the consumer to explicitly act on potential errors. I must be honest here that 

Reviewer: Think the title `Allows you to provide parameter validation as early as possible ` and then "Which" is a bit weird.

> we mostly need this level of validation due to Go's inability to provide non-nil pointer flowing 

Reviewer: What is meant by "non-nil pointer flowing"?

Author: Thanks for spotting this, it was a mistake. I meant nil pointer check enforcement. Fixed it now.

> we mostly need this level of validation due to Go's inability to provide non-nil pointer flowing 

Reviewer 3: I must be honest here, we mostly need this level ...

which is provided in languages like TypeScript. if you check out [my post on TypeScript](https://www.telerik.com/blogs/uncovering-typescript-for-c-developers#no-more-billion-dollar-mistakes-protection-against-null-and-undefined), 

> you can see here on what I mean by this. However, there are still genuine other cases where a 

Reviewer 3: > if you check out [my post on TypeScript](https://www.telerik.com/blogs/uncovering-typescript-for-c-developers#no-more-billion-dollar-mistakes-protection-against-null-and-undefined), 
you can see here on what I mean by this.

I would rephrase it to:
My post on TypeScript demonstrates what I mean by this.

> you can see here on what I mean by this. However, there are still genuine other cases where a 

Reviewer 3: ... However, there are genuinely other cases where a compiler ...

> compiler cannot guard against your own business logic. In the case of our example above, 

Reviewer 3: ... In our example above...

we can still make our code compile successfully with the constructor implementation but 
get a runtime error:

```golang
package main

import (
	"fmt"
	"github.com/tugberkugurlu/go-package-scope/singers"
)

func main() {
	s := singers.NewJazzSinger(nil)
	singToConsole(s)
}

func singToConsole(singer singers.Singer) {
	fmt.Println(singer.Sing())
}
```

> When we run, we would see the below error even if we are able to compile the code successfully:

Reviewer 3: When we run we see the error below - even though the code compiled successfully:

goroutine 1 [running]:
github.com/tugberkugurlu/go-package-scope/singers.jazzSinger.Sing(0x0, 0x0, 0x1010095, 0xc00008e030)
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/singers/jazzsinger.go:31 +0x22
main.singToConsole(0x10d75a0, 0xc00008e030)
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/main.go:14 +0x35
main.main()
	/Users/tugberkugurlu/go/src/github.com/tugberkugurlu/go-package-scope/main.go:10 +0x5c
```

There isn't a solution available in Go as far as I am aware which would allow us to fail for these cases during compilation. However, thanks to the dedicated constructor for this object, we can explicitly signal potential construction errors by returning multiple values from the function call:

```golang
func NewJazzSinger(lyrics LyricsProvider) (Singer, error) {
	if lyrics == nil {
		return nil, errors.New("lyrics cannot be nil")
	}
	return jazzSinger{
		lyrics: lyrics,
	}, nil
}
```

At the time of consumption, it becomes very explicit to deal with returned result:

```golang
s, err := singers.NewJazzSinger(nil)
if err != nil {
	log.Fatal(err)
}
// ...
```

### Allows you to control the flow of your implementation

> The below code is a simplified and intended-use scenario of an interesting bug we had on production a while ago:

Reviewer 3: The code below...

... We had **_in_** production ...

```golang
package main

import (
	"fmt"
)

type JazzSinger struct {
	count int
}

func (j *JazzSinger) Sing() string {
	j.count++
	return "Des yeux qui font baisser les miens"
}

func (j *JazzSinger) Count() int {
	return j.count
}

func main() {
	s := &JazzSinger{}
	singToConsole(s)
	fmt.Println(s.Count())
	singToConsole(s)
	fmt.Println(s.Count())
}

func singToConsole(singer *JazzSinger) {
	fmt.Println(singer.Sing())
}
```

> This code works as expected and the singer sings as expected, and count is incremented as expected, all great!

Reviewer 3: This code works as expected: the singer sings, and the count is incremented. All great!

```
Des yeux qui font baisser les miens
1
Des yeux qui font baisser les miens
2
```

This works because our method signature on the `JazzSinger` struct accepts a pointer to an instance of `JazzSinger` which 
means that the count will be incremented as expected even if the type is passed around, and that's what's happening with 
the above scenario. 

However, can we guess what will happen if we change our usage as below:

```golang
func main() {
	s := JazzSinger{}
	singToConsole(s)
	fmt.Println(s.Count())
	singToConsole(s)
	fmt.Println(s.Count())
}

func singToConsole(singer JazzSinger) {
	fmt.Println(singer.Sing())
}
```

> My first would have been that compiler will fail here, and this is a perfectly reasonable assumption to make 

Author: ```suggestion
My first assumption would have been that compiler will fail here, and this is a perfectly reasonable assumption to make 
```

> My first would have been that compiler will fail here, and this is a perfectly reasonable assumption to make 

Reviewer 3: My first guess ....

> since, you know, we are not passing a pointer to `Sing` method call. Well, if you are making the same the assumption I have made, 

Reviewer 3: I personally don't like the `you know` bit here. "since we are not passing a pointer to the `Sing` method call.

> you will be so wrong. This compiles perfectly but it won't work as expected:

Reviewer 3: ... If you made the same assumption as I did, you would be wrong ...

```
Des yeux qui font baisser les miens
0
Des yeux qui font baisser les miens
0
```

> The worst part is that this will actually work if we were to get rid of the `singToConsole` function and embed its implementation:

Reviewer 3: ... this **_would_** actually work ...

```golang
func main() {
	s := JazzSinger{}
	s.Sing()
	fmt.Println(s.Count())
	s.Sing()
	fmt.Println(s.Count())
}
```

```
Des yeux qui font baisser les miens
1
Des yeux qui font baisser les miens
2
```

This is the exact reason why your tests will pass even if they have the wrong usage!

```golang
package main

import (
	"github.com/deliveroo/assert-go"
	"testing"
)

func TestJazzSinger(t *testing.T) {
	t.Run("count increments as expected", func(t *testing.T) {
		singer := JazzSinger{}
		singer.Sing()
		singer.Sing()
		assert.Equal(t, singer.Count(), 2)
	})
}
```

```
➜  jazz-singer git:(master) ✗ go test -v
=== RUN   TestJazzSinger
=== RUN   TestJazzSinger/count_increments_as_expected
--- PASS: TestJazzSinger (0.00s)
    --- PASS: TestJazzSinger/count_increments_as_expected (0.00s)
PASS
ok  	github.com/tugberkugurlu/algos-go/jazz-singer	0.549s
```

> After a bit more digging, it actually turned out that this is actually the intended behavior of Go, and it's 

Reviewer 3: The word "actually" is a bit overused here:

... it turns out that this actually is the intended behaviour of Go ...

even [documented in its spec](https://golang.org/ref/spec#Calls):

> A method call x.m() is valid if the method set of (the type of) x contains m and the argument list can be assigned to the parameter list of m. If x is addressable and &x's method set contains m, x.m() is shorthand for (&x).m().

> I am still unsure why this could be useful, but it's what it's, and it's so easy to make the same mistake since you can ensure how 

Reviewer: > it's what it's

Not idiomatic, use "it is what it is"

> I am still unsure why this could be useful, but it's what it's, and it's so easy to make the same mistake since you can ensure how 

Author: ```suggestion
I am still unsure why this could be useful, but it is what it is, and it's so easy to make the same mistake since you can ensure how 
```

the consumer will flow the type as the creator of the type if it can be constructed freely. In fact, the decision of how the type 

> should be flowed should be the decision of the owner (i.e. its package) of the type, not the consumer, and I have never found a case 

Reviewer: I don't think I've come across flowing as a concept in any statically typed language - what do you mean by it?

Author: The decision of the how the type if flowed is implicitly being carried by the declaration choice you make, e.g. `struct` vs `class` in C#. it's the author's responsibility to make the choice.

Reviewer 4: > In fact, the decision of how the type should be flowed should be the decision of the owner (i.e. its package) of the type, not the consumer

I'd say "the decision of which semantics to use should belong to the author of the package, not to its users"

type by forcing them to choose between a `class` and `struct`.

In Go, you can make this safer through the use of the constructor pattern as well, by ensuring that your struct is not allowed to be constructed directly and you controlling how the initialized value should be flowed.

```golang
package singers

type Singer interface {
	Sing()  string
	Count() int
}

type jazzSinger struct {
	count int
}

func (j *jazzSinger) Sing() string {
	j.count++
	return "Des yeux qui font baisser les miens"
}

func (j *jazzSinger) Count() string {
	return j.count
}

func NewJazzSinger() Singer {
	return &jazzSinger{}
}
```

The consumer of this type needs to construct it through `NewJazzSinger` function here, which is making the decision to flow the type as a pointer because it needs to be able to mutate its own state as it's being used.

```golang
package main

import (
	"fmt"
	"github.com/tugberkugurlu/go-package-scope/singers"
)

func main() {
	s := singers.NewJazzSinger()
	singToConsole(s)
	fmt.Println(s.Count())
	singToConsole(s)
	fmt.Println(s.Count())
}

func singToConsole(singer singers.Singer) {
	fmt.Println(singer.Sing())
}
```

## Conclusion

> Modeling your domain is hard and it's even harder if you have rich models which hold a mutable state along with explicit behaviours. Go programming language may not give you all the tools directly to model your domain in a rich way as some other programming languages provide. However, it's still possible to 

Reviewer: Modelling

> Modeling your domain is hard and it's even harder if you have rich models which hold a mutable state along with explicit behaviours. Go programming language may not give you all the tools directly to model your domain in a rich way as some other programming languages provide. However, it's still possible to 

Reviewer 3: Go programming language may may not give you all the tools to directly model your domain ...

> Modeling your domain is hard and it's even harder if you have rich models which hold a mutable state along with explicit behaviours. Go programming language may not give you all the tools directly to model your domain in a rich way as some other programming languages provide. However, it's still possible to 

Author: ```suggestion
Modelling your domain is hard and it's even harder if you have rich models which hold a mutable state along with explicit behaviours. Go programming language may not give you all the tools directly to model your domain in a rich way as some other programming languages provide. However, it's still possible to 
```

---
layout: post
title:  "Increase the Reliability of a Go Codebase with Object Constructors"
authors:
  - "Tugberk Ugurlu"
excerpt: >
  One of the limitations of Go programming language is the lack of built-in object constructor 
  support. In this post, we will see how this can have a negative impact on the code we write 
  and how we can get around that by gluing together some of the existing language concepts.
---

Coming from a heavy production experience with languages such as C# and TypeScript, I must admit 
that my journey with Go has been a bumpy ride so far, but it's for sure a positive one overall. 
Go certainly shines in some parts such as its runtime efficiency, built-in 
tooling support and its simplicity which allows you to get up to speed with it so quickly! 
However, there are some areas where it limits your ability to express and model your 
software in code in a robust way, especially in a codebase where you get to work on as a team 

> such as lack of [sum types](https://github.com/golang/go/issues/19412) and generics support (lucky, [generics support seems to be on its way](https://go.googlesource.com/proposal/+/master/design/go2draft-generics-overview.md)). 

Reviewer 4: ```suggestion
such as lack of [sum types](https://github.com/golang/go/issues/19412) and generics support (luckily, [generics support seems to be on its way](https://go.googlesource.com/proposal/+/master/design/go2draft-generics-overview.md)). 
```

from the contract by only exposing what we need. The challenge here is to be able to provide a 
way to construct the implementation. Fortunately, there is a workaround for this in Go: we can 
define an exported function within the package, which has access to the internal implementation, 
but also exposes it through the interface, as shown in the example below:

```golang
package singers

type Singer interface {
	Sing() string
}

type jazzSinger struct {
}

func (jazzSinger) Sing() string {
	return "Des yeux qui font baisser les miens"
}

func NewJazzSinger() Singer {
	return jazzSinger{}
}
```

`NewJazzSinger` function here can be accessed by the package consumer but jazzSinger struct is still hidden.

```golang
package main

import (
	"fmt"
	"github.com/tugberkugurlu/go-package-scope/singers"
)

func main() {
	s := singers.NewJazzSinger()
	singToConsole(s)
}

func singToConsole(singer singers.Singer) {
	fmt.Println(singer.Sing())
}
```

Why is this good and how does this make our code more reliable? Let's go over the main advantages of this 
technique, and how they make our code more reliable.

> ### Changes in the struct's fields would make our code fail in compile time, rather than runtime

Reviewer 4: ```suggestion
### Changes in the struct's fields would make our code fail at compile time, rather than runtime
```

Unlike other languages (such as TypeScript), Go doesn't have a way to enforce assigning fields directly 
(omitted fields default to the zero value, which may not always be what you want) - so the compiler 
would not help us here - we would need to track all updates to the struct's fields manually, which is 
tedious and error prone (specially in large codebases). Best case scenario, the code would be well tested 
and the unit tests would break. Worst case scenario, the code would blow up at Runtime, which would require 

> a rollback to this release. To make matters worse, your application could be happily working without any 

Reviewer 4: Maybe _of_? Otherwise not clear which release...
```suggestion
a rollback of this release. To make matters worse, your application could be happily working without any 
```

crashes, but the its behaviour could be wrong due to the way the implementation might work. This one is 

> the hardest and potentially harmful bugs to catch as it could have a larger impact on your efforts and 

Reviewer 4: Since you're using the "hardest" (and not "hard") I guess you should say "most harmful" as well? Also "this" so "bug"?
```suggestion
the hardest and potentially most harmful bug to catch as it could have a larger impact on your efforts and 
```

```golang
package main

import (
	"fmt"
)

type JazzSinger struct {
	count int
}

func (j *JazzSinger) Sing() string {
	j.count++
	return "Des yeux qui font baisser les miens"
}

func (j *JazzSinger) Count() int {
	return j.count
}

func main() {
	s := &JazzSinger{}
	singToConsole(s)
	fmt.Println(s.Count())
	singToConsole(s)
	fmt.Println(s.Count())
}

func singToConsole(singer *JazzSinger) {
	fmt.Println(singer.Sing())
}
```

This code works as expected: the singer sings, and the count is incremented. All great!

```
Des yeux qui font baisser les miens
1
Des yeux qui font baisser les miens
2
```

This works because our method signature on the `JazzSinger` struct accepts a pointer to an instance of `JazzSinger` which 
means that the count will be incremented as expected even if the type is passed around, and that's what's happening with 
the above scenario. 

However, can we guess what will happen if we change our usage as below:

```golang
func main() {
	s := JazzSinger{}
	singToConsole(s)
	fmt.Println(s.Count())
	singToConsole(s)
	fmt.Println(s.Count())
}

func singToConsole(singer JazzSinger) {
	fmt.Println(singer.Sing())
}
```

> My first guess was that compiler will would here, and this is a perfectly reasonable assumption to make 

Reviewer 5: > My first guess was that compiler will would here

Sounds like something is missing here?
